For jenkins_script, bring parity for the support of some parameters with other modules.

haridsv opened the pull request:
##### SUMMARY
- Add support for url params that most other modules support (includinging jenkins-plugin).
- Bring functionality around `params` and `force_basic_auth` from jenkins-plugin over to get some uniformity.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
- jenkins_script

##### ANSIBLE VERSION
```
ansible 2.3.1.0
  config file =
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Aug 16 2017, 20:02:23) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]
```


##### ADDITIONAL INFORMATION
This change provides a means to use some parameters consistently across all plugins and should in general help simplify the usage.

The `params` parameter is a generic pattern for passing in arbitrary parameters to modules, and is already supported by `jenkins_plugin`. Also, some common parameters to the Ansible framework can be passed in via `params` without any extra effort from the plugin.

haridsv commented:
@hogarthj I separated this out of my previous PR #26564 and request you to review.

haridsv commented:
@hogarthj To bring some context from the old PR and add case for the `params` parameter, take the example of `validate_certs`. This parameter is already supported by `jenkins_plugin` via the common `params` parameter, but since `jenkins_script` didn't support `params`, it had to expose as a top-level parameter. This sort of deviations cause unnecessary confusion and surprises among users (as it did for me).

hogarth-sv commented on lib/ansible/modules/web_infrastructure/jenkins_script.py:
```diff
@@ -142,24 +165,35 @@ def get_crumb(module):
-        module.params['force_basic_auth'] = True
+
+    # Force basic authentication
+    module.params['force_basic_auth'] = True
```
It's not safe to always force authentication. This needs to only be set when there are credentials.

haridsv replied:
@jtyr Any idea why `jenkins_plugin` does it unconditionally?

jtyr replied:
Because by default you don't have Jenkins installation without the need to login to be an admin?

haridsv replied:
The default is fine, but the point is that it is being forced unconditionally so @hogarthj's point is that it should be conditional.

On the other hand, I looked at one one of library code paths that uses this flag and noticed that it is being used only if username is set, so there is problem if username is set and the flag is not set, but the opposite case should be fine. Here is the code:

https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/urls.py#L860

hogarth-sv commented on lib/ansible/modules/web_infrastructure/jenkins_script.py:
```diff
@@ -142,24 +165,35 @@ def get_crumb(module):
+        user=dict(required=False, no_log=True, type="str", default=None),
+        password=dict(required=False, no_log=True, type="str", default=None),
+        timeout=dict(required=False, type="int", default=10),
+        params=dict(type='dict'),
```
This needs to be marked as not required.

haridsv replied:
Got it!

haridsv replied:
I think `False` is already the default as was used in `jenkins_plugin`:

https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/web_infrastructure/jenkins_plugin.py#L739

I will double check.

haridsv replied:
Here is where the default comes from: https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L1512

```
            required = v.get('required', False)
```

hogarth-sv commented on lib/ansible/modules/web_infrastructure/jenkins_script.py:
```diff
@@ -142,24 +165,35 @@ def get_crumb(module):
 
 def main():
 
+    argument_spec = url_argument_spec()
```
This makes sense to switch to ... but for compatibility and simplicity user and password should be made aliases of url_username and url_password ... and if using the framework then the notes should make it clear that all params supported by it are possible.

haridsv replied:
Yes, that makes sense. Looks like `user` and `password` are a couple of other parameters that diverged from `jenkins_plugin`, so I will try to bring some uniformity to this.

hogarth-sv commented on lib/ansible/modules/web_infrastructure/jenkins_script.py:
```diff
@@ -101,6 +108,22 @@
+# Define a variable and specify all default parameters you want to use across
+# all jenkins_plugin calls:
+#
+# my_jenkins_params:
```
This is confusing ... if the example is meant to show creating a variable to be used by the params option then don't comment out that variable.

Also I'm not sure we need to show this specifically ... just the params with a dict hanging off should suffice.

haridsv replied:
This was a direct copy from the way it was documented in jenkins_plugin.py:

https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/web_infrastructure/jenkins_plugin.py#L173

haridsv replied:
There was a recent fix to `jenkins_plugin` to avoid exposing the `url_password` in logs, perhaps it is worth bringing that idea to this plugin as well. Here is the change:

https://github.com/ansible/ansible/commit/863fcb5aced55fd9ce3080905bd6405c8bfd0511#diff-d2e810bf269df1e341b77e8baae7d35e

hogarth-sv reviewed (changes requested):
I can see the use case better now than in the previous PR and I don't have any fundamental objection to the change, but lets avoid breaking the current behaviour for people but at the same time ensure consistency of variables and behaviour.

If we do this let's make user and password a direct alias of url_username and url_password. 

This does have the benefit of automatically supporting all new params functionality added to the fetch_url framework without specific changes required here... which I support.

We do need to clean up the docs to make this clear though and remove redundancy (eg validate_certs) where it exists.

haridsv commented:
@hogarthj I think I addressed all your concerns in the latest change, please take another look.

haridsv commented:
Hold on, I need to fix the documentation issue. I also found a better way to check for parameter exclusivity, so some simplification can be done.

haridsv commented:
Please review!

haridsv commented:
@hogarthj Wondering if you have any more feedback on this, thanks.

hogarth-sv commented:
@haridsv chill a bit ... I'll get to this in due course.

haridsv commented:
Something seems to be wrong with the merge I did online, to have caused these validation failures. I can diff between the latest `devel` and my merged branch and see scores of others diffs. With the merge with `devel`, I should have see just the diff that corresponds to this PR. If I try to merge again from `devel` on my local clone, git says it is all up to date. Also, I get no validation errors if I checkout the latest `devel` and manually copy over my changes from this PR, so github seems to have messed up the prior merge. I will try to update my PR with a fresh push. This may unfortunately trample some of the history here, but I don't know else what to do.

haridsv committed:
Bring parity for the support of some parameters with other modules.

- Add support for url params that most other modules support (includinging jenkins-plugin).
- Bring functionality around `params` and `force_basic_auth` from jenkins-plugin over to get some uniformity.
- Aliasing of the old user and password parameters to the new url_username and url_password parameters. Also disallow url_password via generic params.
- Fixed documentation and simplified mutual exclusivity checks

haridsv commented:
Damn, I accidentally caused this PR to close, let me try to reopen.

haridsv committed:
Merge branch 'devel' into jenkins-script-params-parity

hogarthj commented:
Note that the params on at least on of the other modules was dropped recently due top issues with it.

As a result I think this may be moot at this point as not only won't there be parity, but the reasons for doing so on the other module may be valid here. 

I'll get to more detail on this after verifying soon.

hogarthj commented:
The issue related is #32708

@abadger you made that change in jenkins_plugin ... your thoughts on this?

abadger commented:
The url_username and url_password additions look good.

As a group, the ansible commiters decided that params-style options are not
something we want in any ansible module.  So they were removed from
Jenkins_plugin and yum_repository (the two known cases).  Currently you can
set args to a dict in the playbook to have similar functionality.

Jimi-c has deprecated that for security reasons but it continues to be a
discussion as most of us see this as a feature that users greatly
appreciate, it's just tricky to do it in a way that doesn't trip over
jimi's security concerns.

On Jan 25, 2018 9:37 AM, "ansibot" <notifications@github.com> wrote:

> The test ansible-test sanity --test validate-modules [?
> <https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/validate-modules.html>]
> failed with the following errors:
>
> lib/ansible/modules/web_infrastructure/jenkins_script.py:0:0: E322 "client_cert" is listed in the argument_spec, but not documented in the module
> lib/ansible/modules/web_infrastructure/jenkins_script.py:0:0: E322 "client_key" is listed in the argument_spec, but not documented in the module
> lib/ansible/modules/web_infrastructure/jenkins_script.py:0:0: E322 "force" is listed in the argument_spec, but not documented in the module
> lib/ansible/modules/web_infrastructure/jenkins_script.py:0:0: E322 "force_basic_auth" is listed in the argument_spec, but not documented in the module
> lib/ansible/modules/web_infrastructure/jenkins_script.py:0:0: E322 "http_agent" is listed in the argument_spec, but not documented in the module
> lib/ansible/modules/web_infrastructure/jenkins_script.py:0:0: E322 "thirsty" is listed in the argument_spec, but not documented in the module
> lib/ansible/modules/web_infrastructure/jenkins_script.py:0:0: E322 "use_proxy" is listed in the argument_spec, but not documented in the module
>
> click here for bot help
> <https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/32430#issuecomment-360541545>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMxWvYR8sxKcLERbeoB201ywH2_kY0Mks5tOLvLgaJpZM4QOA9Q>
> .
>

haridsv commented:
This is no longer relevant.
